# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель, ответить на вопросы:**
- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?
- Приведите возможные причины появления пропусков в исходных данных.
- Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
# импорт библиотеки pandas
import pandas as pd

# чтение файла с данными и сохранение в data
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
# получение первых 20 строк таблицы data
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
# получение общей информации о данных в таблице data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
# выявление аномалий в данных в таблице data
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06


- `children` —  количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита

**Вывод:** 
- В таблице 12 столбцов и 21525 строк. Типы данных — `object`, `int`, `float`
- Есть пропущенные значения в двух столбцах: `days_employed` и `total_income`
- Доля пропущенных значений 10,1% - значительная часть данных
- Есть отрицательные значения в столбцах `children` и `days_employed`
- Тип данных в `total_income` поменять на `int`
- Поменять регистр букв в `education`, чтобы выявить не явные дубликаты
- Удалить явные дубликаты



## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
#Проверим датафрейм на пустые значения
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
#заменим пустые значение в данных о доходах на медианные значения по типу занятости
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
#заменим все отрицательные значения трудового стажа положительными
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
#выведим для каждого типа занятости медианное значение трудового стажа в днях.
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
#выведим перечень уникальных значений столбца количества детей
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
#удалим строки с аномальными значениями
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
#убедимся, что артефакты удалены
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
#заменим пустые значение в данных о трудовом стаже на медианные значения по типу занятости
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
#проверим, что все пропуски заполнены
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [39]:
#заменим вещественный тип данных в столбце total_income на целочисленный методом astype
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
#приведем все к нижнему регистру в столбце образования
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
#проверим наличие дубликатов
data.duplicated().sum()

71

In [17]:
#удалим все дубликаты
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
#напишим функцию для категоризации клиентов по уровню дохода
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
#создадим и добавим в новый столбец соответствующую категорию
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
#выведите на экран перечень уникальных целей взятия кредита
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
#напишим функцию для категоризации клиентов по цели взятия кредита
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
#создадим и добавим в новый столбец соответствующую категорию
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Вывод:**

- Заполнили пропуски в столбце `total_income` медианным значением по каждому типу из столбца `income_type`
- Заменили все отрицательные значения в столбце `days_employed` положительными с помощью метода `abs()`
- Удалили строки с аномальными значениями в стлобце `children`
- Заполнили пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`
- Заменили вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`
- Обработали неявные дубликаты в столбце `education`, привели их к нижнему регистру, с помощью метода `str.lower()`
- Данные категоризированы по уровню дохода и по цели взятия кредита. Для этого написали функции `categorize_income()`, `categorize_purpose()` и использовали метод `apply()`

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
#проверим зависимость между столбцами children и debt
children_data_pivot = pd.pivot_table(data, index='children', values='debt', aggfunc='count')
children_data_pivot['%'] = round(pd.pivot_table(data, index='children', values='debt')*100, 1)
children_data_pivot

debt    %
children            
0         14091  7.5
1          4808  9.2
2          2052  9.5
3           330  8.2
4            41  9.8
5             9  0.0

**Вывод:** Кредиторы не имеющие детей, закрывают кредит чаще на 1-2%, чем те, у кого есть один или более детей. 

Следует заметить, что клиентов с 5 детьми в датасете всего 9, это составляет 0.04% от общего числа клиентов. Данные по данной категории клиентов могут быть нерелевантными.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
#проверим зависимость между столбцами family_status и debt
family_status_data_pivot = pd.pivot_table(data,index='family_status', values='debt', aggfunc='count')
family_status_data_pivot['%'] = round(pd.pivot_table(data,index='family_status', values='debt')*100, 1)
family_status_data_pivot.sort_values(by='%',ascending=True)

debt    %
family_status                    
вдовец / вдова           951  6.6
в разводе               1189  7.1
женат / замужем        12261  7.6
гражданский брак        4134  9.3
Не женат / не замужем   2796  9.8

**Вывод:** Семейное положение положительно влияет на погашение кредита в срок. Количество клиентов, у которых был или есть официально зарегистрированный брак и которые имеют задолженность по кредиту, ниже на 2-3%

- `вдовец / вдова` чаще всех погашают кредит своевременно — 6,6%. 
- `Не женатые / не замужние` чаще всех имеют задолженности с погашением кредита — 9,8%.
- разница между наибольшим и наименьшим значением составляет ~3%.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [38]:
#проверим зависимость между столбцами total_income_category и debt

income_data_pivot = data.groupby('total_income_category').agg({'debt':['sum', 'count', 'mean']})

#income_data_pivot = pd.pivot_table(data, index='total_income_category', values='debt', aggfunc='count')
#income_data_pivot['%'] = round(pd.pivot_table(data, index='total_income_category', values='debt') * 100, 1)

income_data_pivot


debt                 
                        sum  count      mean
total_income_category                       
A                         2     25  0.080000
B                       354   5014  0.070602
C                      1353  15921  0.084982
D                        21    349  0.060172
E                         2     22  0.090909

**Вывод:**  Внутри последовательности не прослеживается корреляции между уровнем дохода и процентом задолженностей. Можно сделать вывод, что зависимости нет.

- Люди из категории`E` (0-30000) чаще всех имеют задолженности с погашением кредита — 9,1%
- Люди из категории `D` (30001-50000) чаще всех погашают кредит своевременно — 6%
- Разница между наибольшим и наименьшим значением составляет ~3%.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
#проверим зависимость между столбцами purpose_category и debt
purpose_data_pivot = pd.pivot_table(data, index='purpose_category', values='debt', aggfunc='count')
purpose_data_pivot['%'] = round(pd.pivot_table(data, index='purpose_category', values='debt') * 100, 2)
purpose_data_pivot.sort_values(by='%',ascending=True)

debt     %
purpose_category                     
операции с недвижимостью  10751  7.26
проведение свадьбы         2313  7.91
получение образования      3988  9.25
операции с автомобилем     4279  9.35

**Вывод:** 
- Люди, взявшие кредит на недвижимость, чаще своевременно погашают кредит — 7,3%
- Чаще всех имеют задолженность с погашением кредита на покупку автомобиля — 9,3% 
- Разница между наибольшим и наименьшим значением составляет ~2%


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
- Человек не указал при заполнении анкеты
- Сбой при выгрузке данных, такие пропуски можно узнать у разработчиков
- Человек никогда не работал, поэтому стажа и дохода у него нет


#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медианное значение для количественных данных является лучшим решением для заполнения пропусков, т.к. исключает в расчете экстремальные значения. Ведь когда есть оторванные от основной массы лидеры, например зарабатывающие значительно больше, их результаты сильно смещают значение среднего вверх.

### Шаг 4: общий вывод.

Семейное положение, наличие детей, уровень дохода и цель кредита влияет на факт погашения кредита в срок.

- лица без детей, со средним или выше среднего доходом, берущие кредит с целью улучшения жилищных условий или на другие операции с недвижимостью, являются наиболее надежными заёмщиками;

- лица, не состоящие в официальном браке, но имеющие детей, с низким доходом, желающие потратить заёмные средства на покупку автомобиля, наиболее склонны к нарушению договорных обязательств и погашению кредита в срок.


В ходе обзора данных выявили, что:
- В таблице 12 столбцов и 21525 строк. Типы данных — `object`, `int`, `float`
- Есть пропущенные значения в двух столбцах: `days_employed` и `total_income`
- Доля пропущенных значений 10,1% - значительная часть данных
- Есть отрицательные значения в столбцах `children` и `days_employed`
- Тип данных в `total_income` нужно поменять на целочисленный `int`
- В столбце `education`, есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв


В ходе предобработки данных:
- Заполнили пропуски в столбце `total_income` медианным значением по каждому типу из столбца `income_type`
- Заменили все отрицательные значения в столбце `days_employed` положительными
- Удалили строки с аномальными значениями в стлобце `children`
- Заполнили пропуски в столбце `days_employed` медианными значениями по каждому типу занятости `income_type`
- Заменили вещественный тип данных в столбце `total_income` на целочисленный
- Обработали неявные дубликаты в столбце `education`
- Данные категоризировали по уровню дохода и по цели взятия кредита

### Чек-лист готовности проекта

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.
